cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java

```diff
@@ -45,6 +46,43 @@ public WorkflowSpecification(String className, String name, String description,
     this.properties = properties == null ? Collections.<String, String>emptyMap() :
                                            Collections.unmodifiableMap(new HashMap<>(properties));
     this.nodes = Collections.unmodifiableList(new ArrayList<>(nodes));
+    this.nodeIdMap = Collections.unmodifiableMap(generateNodeIdMap());
+  }
+
+  /**
+   * Visit all the nodes in the {@link Workflow} and generate the map of node id to
+   * the {@link WorkflowNode}. Returned map does not contain any mapping for the FORK node,
+   * however contains the ACTION and CONDITION nodes within the FORK node itself.
+   */
+  private Map<String, WorkflowNode> generateNodeIdMap() {
+    Map<String, WorkflowNode> nodeIdMap = new HashMap<>();
+    Queue<WorkflowNode> nodes = new LinkedList<>(this.nodes);
```

Comment:
Don't use `this` since this method is called from constructor. Have the List of nodes passed to this method.